File: Algorithm/BOJ/Go/Binary Search/2805.go
// 2805
package main

import (
	"bufio"
	"fmt"
	"os"
)

func main() {
	reader := bufio.NewReader(os.Stdin)
	var N, M int
	fmt.Fscanln(reader, &N, &M)

	trees := make([]int, N)
	for i := 0; i < N; i++ {
		fmt.Fscanf(reader, "%d ", &trees[i])
	}

	fmt.Println(FindMaxHeight(&trees, M))
}

func GetMaxElement(array *[]int) int {
	max := -1

	for _, element := range *array {
		if element > max {
			max = element
		}
	}
	return max
}

func FindMaxHeight(trees *[]int, target_log int) int {
	left := 1
	right := GetMaxElement(trees)
	var mid int

	for left <= right {
		mid = (left + right) / 2
		log := 0

		for _, tree := range *trees {
			if tree > mid {
				log += (tree - mid)
			}
		}

		if log >= target_log {
			left = mid + 1
		} else {
			right = mid - 1
		}
	}
	return right
}

File: Algorithm/BOJ/C++/Binary Search/2805.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

long FindMaxHeight(vector<int>* trees, int target_log);

int main() {
	int N, M, i, tree;

	cin >> N >> M;
	vector<int> trees;

	for (i = 0; i < N; i++) {
		cin >> tree;
		trees.push_back(tree);
	}

	cout << FindMaxHeight(&trees, M);
}

long FindMaxHeight(vector<int>* trees, int target_log) {
	int left = 1;
	int right = *max_element(trees->begin(), trees->end());
	int mid;

	while (left <= right) {
		mid = (left + right) / 2;
		long log_sum = 0;

		for (int tree : *trees) {
			if (tree > mid) {
				log_sum += tree - mid;
			}
		}

		if (log_sum >= target_log) {
			left = mid + 1;
		}
		else {
			right = mid - 1;
		}
	}
	return right;
}